File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
'''
문제 : 단지 번호
링크 : https://www.acmicpc.net/problem/2667
소요시간 : 30분
'''
n = int(input())
graph = [list(map(int, input())) for _ in range(n)]
apart = []
dx, dy = [-1, 1, 0, 0], [0, 0, -1, 1]
def dfs(r, c, cnt):
    graph[r][c] = 0
    if r < 0 or r >= n or c < 0 or c >= n:
        return cnt
    
    for i in range(4):
        nx, ny = r + dx[i], c + dy[i]
        # 범위 밖은 탐색 제외
        if nx < 0 or nx >= n or ny < 0 or ny >= n:
            continue
        if graph[nx][ny] == 1:
            cnt = dfs(nx, ny, cnt+1)
    return cnt
for i in range(n):
    for j in range(n):
        if graph[i][j] == 1:
            apart.append(dfs(i, j, 1))
            
apart.sort()
print(len(apart))
for i in apart:
    print(i)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.util.*;
import java.io.*;

class Main {
    static int[][] graph;
    static List<Integer> list = new ArrayList<>();
    static int[] dr = {1, -1, 0, 0};
    static int[] dc = {0, 0, -1, 1};
    static int N;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        N = Integer.parseInt(br.readLine());
        graph = new int[N][N];

        for(int i = 0; i < N; i++){
            String str = br.readLine();
            for(int j = 0; j < N; j++){
                graph[i][j] = Character.getNumericValue(str.charAt(j));
            }
        }

        for(int i = 0; i < N; i++){
            for(int j = 0; j < N; j++){
                if(graph[i][j] == 1){
                    list.add(dfs(i, j, 1));
                }
            }
        }

        sb.append(list.size());
        sb.append("\n");
        Collections.sort(list);
        for(int num : list){
            sb.append(num);
            sb.append("\n");
        }
        System.out.println(sb.toString());
    }

    public static int dfs(int r, int c, int cnt){
        graph[r][c] = 0;
        if(r < 0 || r >= N || c < 0 || c >= N)
            return cnt;

        for(int i = 0; i < 4; i++){
            int nr = r + dr[i];
            int nc = c + dc[i];
            // 범위 밖은 탐색 제외
            if(nr < 0 || nr >= N || nc < 0 || nc >= N)
                continue;
            if(graph[nr][nc] == 1)
                cnt = dfs(nr, nc, cnt+1);
        }

        return cnt;
    }
}